README
#$Id$

Net::RTorrent
========================

Net::RTorrent - Perl interface to rtorrent via XML-RPC.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires  Perl 5.005 or greater and libraries:

* RPC::XML
* Collection


AUTHOR

Zahatski Aliaksandr
zag@cpan.org

COPYRIGHT AND LICENCE


Copyright (C) 2008 Zahatski Aliaksandr

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for Perl extension Net::RTorrent.
#$Id$

0.11 Apr 5, 2011
    - fix CPAN distr
    
0.10 Oct 26, 2010
    - fix make instance on socket at RTorrent::Socket. ( submitted by  James Nylen )

0.09 05.07.2010
    - add use directive  RPC::XML::Parser;
    - comment buggy fields ( thanks Duff aka cmacduff@)

0.08 05.07.2010 
    -fix tests

0.07
    - add is_active and is_open
0.06
    - add direct connect to rtorrent via INET or UNIX socket
0.05
    - fix Collection
0.04
    - add d.get_custom2 support
0.03
    - handle <detete> operation; clean torrent and downloaded files
0.02
    - add system_stat
    - add load_raw

0.01
    - create classes


lib/Net/RTorrent.pm
#$Id$

package Net::RTorrent;

use strict;
use warnings;
use RPC::XML;
use RPC::XML::Client;
use Net::RTorrent::Downloads;
use Net::RTorrent::Socket;
use Collection;
our @ISA     = ();
use Carp;
use 5.005;

=head1 NAME

Net::RTorrent - Perl interface to rtorrent via XML-RPC.

=head1 SYNOPSIS

  #from http scgi gate
  my $obj =  new Net::RTorrent:: 'http://10.100.0.1:8080/scgitest';
  #from network address
  my $obj =  new Net::RTorrent:: '10.100.0.1:5000';
  #from UNIX socket
  my $obj =  new Net::RTorrent:: '/tmp/rtorrent.sock';
  
  #get completed torrents list
  my $dloads = $obj->get_downloads('complete');
  #get all torrents list
  my $dloads = $obj->get_downloads();
  #get stopped torrents list
  my $dloads = $obj->get_downloads('stopped');
  
  #fetch all items
  $dloads->fetch()
  #or by hash_info
  $dloads->fetch_one('02DE69B09364A355F71279FC8825ADB0AC8C3A29')
  #list oll hash_info
  my $keys = $dloads->list_ids;
  #upload remotely
  $obj->create( $torrent_raw );
  $obj->create( $data, 0 );

=head1 ABSTRACT
 
Perl interface to rtorrent via XML-RPC

=head1 DESCRIPTION

Net::RTorrent - short way to create tools for rtorrent.

=cut

use constant {
    #info atributes for system info
    S_ATTRIBUTES => [
        'get_download_rate'    => 'download_rate',    #in my version dosn't work
        'get_memory_usage'     => 'memory_usage',
        'get_max_memory_usage' => 'max_memory_usage',
        'get_name'             => 'name',
        'get_safe_free_diskspace' => 'safe_free_diskspace',
        'get_upload_rate'         => 'upload_rate',
        'system.client_version'   => 'client_version',
        'system.hostname'         => 'hostname',
        'system.library_version'  => 'library_version',
        'system.pid'              => 'pid',
    ],
};

our $VERSION = '0.11';
my $attrs = {
    _cli       => undef,
};
### install get/set accessors for this object.
for my $key ( keys %$attrs ) {
    no strict 'refs';
    *{ __PACKAGE__ . "::$key" } = sub {
        my $self = shift;
        $self->{$key} = $_[0] if @_;
        return $self->{$key};
      }
}

=head1 METHODS

=cut

=head2 new URL

Creates a new client object that will route its requests to the URL provided. 

=cut

sub new {
    my $class = shift;
    $class = ref $class if ref $class;
    my $self = bless( {}, $class );
    if (@_) {
        my $rpc_url = shift;
        my $cli_class  = $rpc_url =~m%\w+://% ? 'RPC::XML::Client': 'Net::RTorrent::Socket';
        $self->_cli( $cli_class->new($rpc_url) );

    }
    else {
        carp "need xmlrpc server URL";
        return;
    }
    return $self;
}

=head2 create \$raw_data || new IO::File , [ start_now=>1||0 ],[ tag=><string>]

Load torrent from file descriptor or scalar ref.

Params:

=over 2 

=item start_now  - start torent now

1 - start download now (default)

0 - not start download

=item tag - save <string> to rtorrent

For read tag value use:

    $ditem->tag

=back

=cut

sub create {
    my $self = shift;
    my $res = $self->load_raw(@_);
    return $res
}


sub load_raw {
    my $self = shift;
    my ( $raw, %flg ) = @_;
    $flg{start_now} = 1 unless defined $flg{start_now};
    my $command = $flg{start_now} ? 'load_raw_start' : 'load_raw';
    my @add =();
    push @add, "d.set_custom2=$flg{tag}" if exists $flg{tag};
    return $self->_cli->send_request( $command, RPC::XML::base64->new($raw), @add );
}


=head2 delete (<info_hash1>[, <info_hash2> ... ])

Call d.erase on I<info_hashes>.

return { <info_hashes> => <xml-rpc response value> }

=cut

sub _delete {
    my $self = shift;
    my (@ids) = map { ref($_) ? $_->{id} : $_ } @_;
    my %res = ();
    for (@ids) {
        my $resp = $self->_cli->send_request( 'd.erase', $_ );
        if ( ref $resp ) {
            $res{$_} = $resp->value;
        }
    }
    return \%res;
}



=head2 list_ids ( [ <name of view> ])

Return list of rtorrent I<info_hashes> for I<name of view>.
An empty string for I<name of view> equals "default".

To get list of views names :

    xmlrpc http://10.100.0.1:8080/scgitest view_list

  'main'
  'default'
  'name'
  'started'
  'stopped'
  'complete'
  'incomplete'
  'hashing'
  'seeding'
  'scheduler'

=cut

sub list_ids {
    my $self = shift;
    my $cli  = $self->_cli;
    my $resp = $cli->send_request('download_list',shift ||"default");
    return ref($resp) ? $resp->value : [];
}


=head2 get_downloads [ <view name > || 'default']

Return collection of downloads (L< Net::RTorrent::Downloads>).

To get list of view:

    xmlrpc http://10.100.0.1:8080/scgitest view_list

  'main'
  'default'
  'name'
  'started'
  'stopped'
  'complete'
  'incomplete'
  'hashing'
  'seeding'
  'scheduler'

=cut

sub get_downloads {
    my $self = shift;
    my $view = shift;
    return new Net::RTorrent::Downloads:: $self->_cli, $view;
}

=head2 system_stat 

Return system stat.

For example:

    print Dumper $obj->system_stat;

Return:

        {
           'library_version' => '0.11.9',
           'max_memory_usage' => '-858993460', #  at my amd64
           'upload_rate' => '0',
           'name' => 'gate.home.zg:1378',
           'memory_usage' => '115867648',
           'download_rate' => '0',
           'hostname' => 'gate.home.zg',
           'pid' => '1378',
           'client_version' => '0.7.9',
           'safe_free_diskspace' => '652738560'
         };

=cut

sub system_stat {
    my $self  = shift;
    my $comms = S_ATTRIBUTES;
    my @list  = @{$comms};
    my ( @res_pull, @cmd_pull ) = ();
    while ( my ( $mname, $aname ) = splice( @list, 0, 2 ) ) {
        push @res_pull, $aname;
        push @cmd_pull, $mname => [];
    }
    my $call_res = $self->do_sys_mutlicall(@cmd_pull);
    my %res      = ();
    while ( my $tmp_res = shift @$call_res ) {
        my $attr_name = shift @res_pull;
        $res{$attr_name} = defined $tmp_res->[1] ? $tmp_res : $tmp_res->[0];
    }
    return \%res

}

=head2 do_sys_mutlicall 'method1' =>[ <param1>, .. ], ...

Do XML::RPC I<system.multicall>. Return ref to ARRAY of results

For sample.

 print Dumper $obj->do_sys_mutlicall('system.pid'=>[], 'system.hostname'=>[]);

Will return:

    [
           [
             '1378'
           ],
           [
             'gate.home.zg'
          ]
    ];

=cut

sub do_sys_mutlicall {
    my $self    = shift;
    my $res     = [];
    my @methods = ();
    while ( my ( $method, $param ) = splice( @_, 0, 2 ) ) {
        push @methods, { methodName => $method, params => $param },;
    }
    if (@methods) {
        my $resp =
          $self->_cli->send_request(
            new RPC::XML::request::( 'system.multicall', \@methods ) );
        $res = $resp->value;
    }
    return $res;
}

1;
__END__

=head1 Setting up rtorrent

If you are compiling from rtorrent's source code, this is done during the configuration step by adding the flag --with-xmlrpc-c to the configure step. Example ./configure --with-xmlrpc-c. See L<http://libtorrent.rakshasa.no/wiki/RTorrentXMLRPCGuide>

Setup your rtorrent  and Web server. My tips:

=head3 .rtorrent

   scgi_port = 10.100.0.1:5000 
   #for complete erase
   on_erase = erase_complete,"execute=rm,-rf,$d.get_base_path="
   #or for save backup 
   on_erase = move_complete,"execute=mv,-n,$d.get_base_path=,~/erased/ ;d.set_directory=~/erased"

=head3 apache.conf

    LoadModule scgi_module        libexec/apache2/mod_scgi.so
    <IfModule  mod_scgi.c>
      SCGIMount /scgitest 10.100.0.1:5000
      <Location "/scgitest">
         SCGIHandler On
      </Location>
    </IfModule>

My url for XML::RPC is L<http://10.100.0.1:8080/scgitest>.

Use B<xmlrpc> ( L<http://xmlrpc-c.sourceforge.net/> ) for tests:

    xmlrpc http://10.100.0.1:8080/scgitest system.listMethods


=head1 SEE ALSO

Net::RTorrent::DItem, Net::RTorrent::Downloads, L<http://libtorrent.rakshasa.no/wiki/RTorrentXMLRPCGuide>

=head1 AUTHOR

Zahatski Aliaksandr, E<lt>zag@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2008-2009 by Zahatski Aliaksandr

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


lib/Net/RTorrent/DItem.pm
#$Id$

package Net::RTorrent::DItem;

=head1 NAME

Net::RTorrent::DItem - Class of rtorrent item.

=head1 SYNOPSIS

  my $obj =  new Net::RTorrent:: 'http://10.100.0.1:8080/scgitest';
  my $dloads = $obj->get_downloads('complete');
  my $keys = $dloads->list_ids;
  my $ditem = $dloads->get_one( $key);
  if ( $ditem->attr->{left_bytes} ) {
    $ditem->stop
  }


=head1 ABSTRACT
 
Net::RTorrent::DItem - Class of rtorrent item.

=head1 DESCRIPTION

Net::RTorrent::DItem - Class of rtorrent item.

=cut

use strict;
use warnings;
use Carp;
use Data::Dumper;
use Collection::Utl::Base;
use Collection::Utl::Item;
use 5.005;
__PACKAGE__->attributes(qw / _cli/);
our @ISA     = qw(Collection::Utl::Item);
our $VERSION = '0.01';
sub _changed { return 0 }

=head1 METHODS

=cut

=head2  attr

Attributes for torrent.

=cut

sub attr {
    return $_[0]->_attr;
}

sub init {
    my $self = shift;
    $self->_cli(shift);
    return $self->SUPER::init(@_);
}

=head2 stop

Stop torrent

=cut

sub stop {
    my $self = shift;
    $self->_cli->send_request();
}

1;
__END__

=head1 SEE ALSO

Collection::Utl::Item

=head1 AUTHOR

Zahatski Aliaksandr, E<lt>zag@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2008 by Zahatski Aliaksandr

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut


lib/Net/RTorrent/Socket.pm
#$Id$

package Net::RTorrent::Socket;

use strict;
use warnings;
use RPC::XML;
use RPC::XML::Client;
use IO::Socket::INET;
use IO::Socket::UNIX;
use RPC::XML::Parser;
use Carp;
use 5.005;

=head1 NAME

Net::RTorrent::Socket - Direct connect to rtorrent via scgi proto

=head1 SYNOPSIS

  my $scli1 = new Net::RTorrent::Socket:: 'localhost:5000';
  my $req = RPC::XML::request->new('get_memory_usage');
  my $res = $scli3->send_request($req);
  print $res->value


=head1 ABSTRACT
 
Perl interface to rtorrent via scgi

=head1 DESCRIPTION

Net::RTorrent::Socket - Direct connect to rtorrent via scgi proto

=cut

our @ISA     = qw();
our $VERSION = '0.06';

sub new {
    my $class = shift;
    $class = ref $class if ref $class;
    my $self = bless( {}, $class );
    if (@_) {
        my $rpc_addr = shift;
        $self->{addr} = $rpc_addr;
    }
    else {
        carp "need xmlrpc server URL";
        return;
    }
    return $self;
}

sub _create_socket {
    my $self   = shift;
    my $addr   = shift || $self->{addr};
    my $type   = $addr =~ m%/% ? "IO::Socket::UNIX" : "IO::Socket::INET";
    my $socket = $type->new( $addr );
    my $old_fh = select($socket);
    $| = 1;
    select($old_fh);
    return $socket

}

sub send_request {
    my ( $self, $req, @args ) = @_;

    if ( !UNIVERSAL::isa( $req, 'RPC::XML::request' ) ) {

        # Assume that $req is the name of the routine to be called
        $req = RPC::XML::request->new( $req, @args );
        return "Error creating RPC::XML::request object: $RPC::XML::ERROR"
          unless ($req);    # $RPC::XML::ERROR is already set
    }

    # get socket
    my $socket = $self->_create_socket( $self->{addr} );
    my $len  = $req->length;
    my $oheaders =
        "CONTENT_LENGTH\0$len\0" 
      . "SCGI\0" . '1' . "\0"
      . "REQUEST_METHOD\0POST\0";
    print $socket length($oheaders) . ":" . $oheaders . "," . $req->as_string;
    my $ans = '';
    while (<$socket>) {
        $ans .= $_;
    }
    $socket->close;

    my ( $header, $xml ) = split( /\n\s?\n/, $ans );
    my $p = RPC::XML::Parser->new();
    my $value;

    eval { $value = $p->parse($xml) };
    if ($@) {
        return $@
    }
    $value->value;
}
1;


lib/Net/RTorrent/Downloads.pm
#$Id$

package Net::RTorrent::Downloads;

use strict;
use warnings;
use Carp;
use Data::Dumper;
use Collection;
use Net::RTorrent::DItem;
use constant {
    D_ATTRIBUTES => [
        'd.get_hash='               => 'hash',
        'd.get_base_filename='      => 'base_filename',
        'd.get_base_path='          => 'base_path',
        'd.get_bytes_done='         => 'bytes_done',
        'd.get_chunk_size='         => 'chunk_size',
        'd.get_chunks_hashed='      => 'chunks_hashed',
        'd.get_complete='           => 'complete',
        'd.get_completed_bytes='    => 'completed_bytes',
        'd.get_completed_chunks='   => 'completed_chunks',
        'd.get_connection_current=' => 'connection_current',
        'd.get_connection_leech='   => 'connection_leech',
        'd.get_connection_seed='    => 'connection_seed',
        'd.get_creation_date='      => 'creation_date',

        'd.get_directory='           => 'directory',
        'd.get_down_rate='           => 'down_rate',
        'd.get_down_total='          => 'down_total',
        'd.get_free_diskspace='      => 'free_diskspace',
        'd.get_hashing='             => 'hashing',
        'd.get_ignore_commands='     => 'ignore_commands',
        'd.get_left_bytes='          => 'left_bytes',
        'd.get_local_id='            => 'local_id',
        'd.get_local_id_html='       => 'local_id_html',
        'd.get_max_file_size='       => 'max_file_size',
#        'd.get_max_peers='           => 'max_peers', 
        'd.get_max_size_pex='        => 'max_size_pex',
#        'd.get_max_uploads='         => 'max_uploads',
        'd.get_message='             => 'message',
#        'd.get_min_peers='           => 'min_peers',
        'd.get_name='                => 'name',
        'd.get_peer_exchange='       => 'peer_exchange',
        'd.get_peers_accounted='     => 'peers_accounted',
        'd.get_peers_complete='      => 'peers_complete',
        'd.get_peers_connected='     => 'peers_connected',
        'd.get_peers_max='           => 'peers_max',
        'd.get_peers_min='           => 'peers_min',
        'd.get_peers_not_connected=' => 'peers_not_connected',
        'd.get_priority='            => 'priority',
        'd.get_priority_str='        => 'priority_str',
        'd.get_ratio='               => 'ratio',
        'd.get_size_bytes='          => 'size_bytes',
        'd.get_size_chunks='         => 'size_chunks',
        'd.get_size_files='          => 'size_files',
        'd.get_size_pex='            => 'size_pex',
        'd.get_skip_rate='           => 'skip_rate',
        'd.get_skip_total='          => 'skip_total',
        'd.get_state='               => 'state',
        'd.get_state_changed='       => 'state_changed',
        'd.get_tied_to_file='        => 'tied_to_file',
        'd.get_tracker_focus='       => 'tracker_focus',
        'd.get_tracker_numwant='     => 'tracker_numwant',
        'd.get_tracker_size='        => 'tracker_size',
        'd.get_up_rate='             => 'up_rate',
        'd.get_up_total='            => 'up_total',
        'd.get_uploads_max='         => 'uploads_max',
        'd.get_custom2='             => 'custom2',
        'd.is_active='               => 'is_active',
        'd.is_open='                 => 'is_open'
    ]
};

use 5.005;
our @ISA     = qw(Collection);
our $VERSION = '0.01';
my $attrs = { _cli => undef, _view => 'default' };
### install get/set accessors for this object.
for my $key ( keys %$attrs ) {
    no strict 'refs';
    *{ __PACKAGE__ . "::$key" } = sub {
        my $self = shift;
        $self->{$key} = $_[0] if @_;
        return $self->{$key};
      }
}

sub _init {
    my $self = shift;
    $self->_cli(shift);
    if ( my $view_name = shift ) {
        $self->_view($view_name);
    }
    return $self->SUPER::_init(@_);
}

=head2 delete (<info_hash1>[, <info_hash2> ... ])

Call d.erase on I<info_hashes>.

return { <info_hashes> => <xml-rpc response value> }

=cut

sub _delete {
    my $self = shift;
    my (@ids) = map { ref($_) ? $_->{id} : $_ } @_;
    my %res = ();
    for (@ids) {
        my $resp = $self->_cli->send_request( 'd.erase', $_ );
        if ( ref $resp ) {
            $res{$_} = $resp->value;
        }
    }
    return \%res;
}

sub _fetch {
    my $self = shift;
    my @ids =  @_;
    my ( @methods, @attr_names ) = ();
    my $r         = D_ATTRIBUTES;
    my @meth2attr = @{$r};
    while ( my ( $method, $attr_name ) = splice @meth2attr, 0, 2 ) {
        push @methods,    $method;
        push @attr_names, $attr_name;
    }
    my $cli  = $self->_cli;
    my $resp =
      $cli->send_request( 'd.multicall', $self->_view || "default", @methods );
    if ( ref $resp ) {
        my $res    = $resp->value;
        my %result = ();
        foreach my $rec (@$res) {
            my %attr;
            @attr{@attr_names} = @$rec;
            my $id = $attr{hash};
            $result{$id} = \%attr;
        }
        return \%result;
    }
    return {};
}

=head2 fetch_one <info_hash>

Return L<Net::RTorrent::DItem> object for given I<info_hash>.

Or undef if I<info_hash> not exists.

=cut

sub fetch_one {
    my $self = shift;
    my ($obj) = values %{ $self->fetch(@_) };
    $obj;
}

=head2 fetch <info_hash1>[, <info_hash2> ... ]

Return info about torrents for given I<info_hashes>.

Result: ref to hash of L<Net::RTorrent::DItem> objects

=cut

sub fetch {
    my $self = shift;
    my @ids  = @_;
    unless ( scalar @_ ) {
       @ids = @{$self->list_ids}
    }
    my $res  = $self->SUPER::fetch(@ids);
    my %res  = ();
    foreach my $key (@ids) {
        $res{$key} = $res->{$key} if exists $res->{$key};
    }
    return \%res;
}


sub list_ids {
    my $self = shift;
    my $cli  = $self->_cli;
    my $resp = $cli->send_request('download_list',shift || $self->_view ||  "default");
    return ref($resp) ? $resp->value : [];
}

=head2 start <info_hash1>[, <info_hash2> ... ]

Start torrents

=cut

sub start {
    my $self = shift;
    my $cli  = $self->_cli;
    my %res = ();
    for (@_) {
        my $resp = $self->_cli->send_request( 'd.start', $_ );
        if ( ref $resp ) {
            $res{$_} = $resp->value;
        }
    }
    return \%res;
}

=head2 stop <info_hash1>[, <info_hash2> ... ]

Stop torrents

=cut

sub stop {
    my $self = shift;
    my $cli  = $self->_cli;
    my %res = ();
    for (@_) {
        my $resp = $self->_cli->send_request( 'd.stop', $_ );
        if ( ref $resp ) {
            $res{$_} = $resp->value;
        }
    }
    return \%res;
}

sub _prepare_record {
    my $self = shift;
    my ( $key, $ref ) = @_;
    return new Net::RTorrent::DItem:: $ref, $self->_cli;
}

1;
__END__

=head1 NAME

Net::RTorrent::Downloads - collection of downloads

=head1 SYNOPSIS

  my $obj =  new Net::RTorrent:: 'http://10.100.0.1:8080/scgitest';
  my $dloads = $obj->get_downloads('');

=head1 ABSTRACT
 
Net::RTorrent::Downloads - collection of downloads


=head1 DESCRIPTION

Net::RTorrent::Downloads - collection of downloads

=head1 SEE ALSO

Collections

=head1 AUTHOR

Zahatski Aliaksandr, E<lt>zag@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2008-2009 by Zahatski Aliaksandr

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
